File: verilog/dv/cocotb/user_proj_tests/multi_uart_test/multi_uart_test.c
// SPDX-FileCopyrightText: 2024 ChipFoundry

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//      http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// SPDX-License-Identifier: Apache-2.0

#include <firmware_apis.h>
#include <CF_UART.h>
#include <CF_UART.c>

// UART base addresses for 8 UARTs (64KB spacing to avoid overlap with 16-bit register offsets)
// CF_UART uses 16-bit register addresses up to 0xFFFF (e.g. 0xFF10), so each UART must be 64KB aligned
// within the user project's Wishbone window.
// UARTn @ 0x300n0000
#define UART0_BASE 0x30000000
#define UART1_BASE 0x30010000
#define UART2_BASE 0x30020000
#define UART3_BASE 0x30030000
#define UART4_BASE 0x30040000
#define UART5_BASE 0x30050000
#define UART6_BASE 0x30060000
#define UART7_BASE 0x30070000

#define UART0 ((EF_UART_TYPE_PTR)UART0_BASE)
#define UART1 ((EF_UART_TYPE_PTR)UART1_BASE)
#define UART2 ((EF_UART_TYPE_PTR)UART2_BASE)
#define UART3 ((EF_UART_TYPE_PTR)UART3_BASE)
#define UART4 ((EF_UART_TYPE_PTR)UART4_BASE)
#define UART5 ((EF_UART_TYPE_PTR)UART5_BASE)
#define UART6 ((EF_UART_TYPE_PTR)UART6_BASE)
#define UART7 ((EF_UART_TYPE_PTR)UART7_BASE)

void main(){
    // Enable management gpio as output to use as indicator for finishing configuration  
    ManagmentGpio_outputEnable();
    ManagmentGpio_write(0);
    enableHkSpi(0); // disable housekeeping spi
    
    // Configure GPIO pins for 8 UARTs
    // UART0: TX=1, RX=0
    GPIOs_configure(1, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(0, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    // UART1: TX=3, RX=2
    GPIOs_configure(3, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(2, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    // UART2: TX=5, RX=4
    GPIOs_configure(5, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(4, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    // UART3: TX=7, RX=6
    GPIOs_configure(7, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(6, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    // UART4: TX=10, RX=9
    GPIOs_configure(10, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(9, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    // UART5: TX=12, RX=11
    GPIOs_configure(12, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(11, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    // UART6: TX=14, RX=13
    GPIOs_configure(14, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(13, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    // UART7: TX=16, RX=15
    GPIOs_configure(16, GPIO_MODE_USER_STD_OUTPUT);
    GPIOs_configure(15, GPIO_MODE_USER_STD_INPUT_PULLUP);
    
    GPIOs_loadConfigs(); // load the configuration 
    User_enableIF(); // this necessary when reading or writing between wishbone and user project if interface isn't enabled no ack would be recieve and the command will be stuck
    
    ManagmentGpio_write(1);
    
    // Initialize all UARTs
    EF_UART_enable(UART0);
    EF_UART_enable(UART1);
    EF_UART_enable(UART2);
    EF_UART_enable(UART3);
    EF_UART_enable(UART4);
    EF_UART_enable(UART5);
    EF_UART_enable(UART6);
    EF_UART_enable(UART7);
    // Ensure UART clocks are enabled (required for TX/RX activity)
    EF_UART_setGclkEnable(UART0, 1);
    EF_UART_setGclkEnable(UART1, 1);
    EF_UART_setGclkEnable(UART2, 1);
    EF_UART_setGclkEnable(UART3, 1);
    EF_UART_setGclkEnable(UART4, 1);
    EF_UART_setGclkEnable(UART5, 1);
    EF_UART_setGclkEnable(UART6, 1);
    EF_UART_setGclkEnable(UART7, 1);
    
    // Set TX FIFO thresholds
    EF_UART_setTxFIFOThreshold(UART0, 3);
    EF_UART_setTxFIFOThreshold(UART1, 3);
    EF_UART_setTxFIFOThreshold(UART2, 3);
    EF_UART_setTxFIFOThreshold(UART3, 3);
    EF_UART_setTxFIFOThreshold(UART4, 3);
    EF_UART_setTxFIFOThreshold(UART5, 3);
    EF_UART_setTxFIFOThreshold(UART6, 3);
    EF_UART_setTxFIFOThreshold(UART7, 3);
    
    // Enable TX for all UARTs
    EF_UART_enableTx(UART0);
    EF_UART_enableTx(UART1);
    EF_UART_enableTx(UART2);
    EF_UART_enableTx(UART3);
    EF_UART_enableTx(UART4);
    EF_UART_enableTx(UART5);
    EF_UART_enableTx(UART6);
    EF_UART_enableTx(UART7);
    
    // Send different messages on each UART with delays between them
    // UART0: "Hello\n"
    EF_UART_writeChar(UART0, 'H');
    EF_UART_writeChar(UART0, 'e');
    EF_UART_writeChar(UART0, 'l');
    EF_UART_writeChar(UART0, 'l');
    EF_UART_writeChar(UART0, 'o');
    EF_UART_writeChar(UART0, '\n');
    
    
    // UART1: "World\n"
    EF_UART_writeChar(UART1, 'W');
    EF_UART_writeChar(UART1, 'o');
    EF_UART_writeChar(UART1, 'r');
    EF_UART_writeChar(UART1, 'l');
    EF_UART_writeChar(UART1, 'd');
    EF_UART_writeChar(UART1, '\n');
    
    
    // UART2: "Test2\n"
    EF_UART_writeChar(UART2, 'T');
    EF_UART_writeChar(UART2, 'e');
    EF_UART_writeChar(UART2, 's');
    EF_UART_writeChar(UART2, 't');
    EF_UART_writeChar(UART2, '2');
    EF_UART_writeChar(UART2, '\n');
    
    
    // UART3: "Test3\n"
    EF_UART_writeChar(UART3, 'T');
    EF_UART_writeChar(UART3, 'e');
    EF_UART_writeChar(UART3, 's');
    EF_UART_writeChar(UART3, 't');
    EF_UART_writeChar(UART3, '3');
    EF_UART_writeChar(UART3, '\n');
    
    
    // UART4: "Test4\n"
    EF_UART_writeChar(UART4, 'T');
    EF_UART_writeChar(UART4, 'e');
    EF_UART_writeChar(UART4, 's');
    EF_UART_writeChar(UART4, 't');
    EF_UART_writeChar(UART4, '4');
    EF_UART_writeChar(UART4, '\n');
    
    
    // UART5: "Test5\n"
    EF_UART_writeChar(UART5, 'T');
    EF_UART_writeChar(UART5, 'e');
    EF_UART_writeChar(UART5, 's');
    EF_UART_writeChar(UART5, 't');
    EF_UART_writeChar(UART5, '5');
    EF_UART_writeChar(UART5, '\n');
    
    
    // UART6: "Test6\n"
    EF_UART_writeChar(UART6, 'T');
    EF_UART_writeChar(UART6, 'e');
    EF_UART_writeChar(UART6, 's');
    EF_UART_writeChar(UART6, 't');
    EF_UART_writeChar(UART6, '6');
    EF_UART_writeChar(UART6, '\n');
    
    
    // UART7: "Test7\n"
    EF_UART_writeChar(UART7, 'T');
    EF_UART_writeChar(UART7, 'e');
    EF_UART_writeChar(UART7, 's');
    EF_UART_writeChar(UART7, 't');
    EF_UART_writeChar(UART7, '7');
    EF_UART_writeChar(UART7, '\n');
} 

File: verilog/dv/cocotb/user_proj_tests/multi_uart_test/multi_uart_test.py
# SPDX-FileCopyrightText: 2024 ChipFoundry

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#      http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# SPDX-License-Identifier: Apache-2.0


from caravel_cocotb.caravel_interfaces import test_configure
from caravel_cocotb.caravel_interfaces import report_test
from caravel_cocotb.caravel_interfaces import UART
import cocotb

@cocotb.test()
@report_test
async def multi_uart_test(dut):
    caravelEnv = await test_configure(dut,timeout_cycles=10000000)
    # wait for start of sending
    await caravelEnv.release_csb()
    await caravelEnv.wait_mgmt_gpio(1)
    
    # Configure UART pins for 8 UARTs
    uart_pins = [
        {"tx": 1, "rx": 0},   # UART0
        {"tx": 3, "rx": 2},   # UART1
        {"tx": 5, "rx": 4},   # UART2
        {"tx": 7, "rx": 6},   # UART3
        {"tx": 10, "rx": 9},  # UART4
        {"tx": 12, "rx": 11}, # UART5
        {"tx": 14, "rx": 13}, # UART6
        {"tx": 16, "rx": 15}  # UART7
    ]
    
    # Create UART instances
    uarts = []
    for i, pins in enumerate(uart_pins):
        uart = UART(caravelEnv, pins)
        uart.bit_time_ns = 200
        uarts.append(uart)
        cocotb.log.info(f"[TEST] Created UART{i} with TX={pins['tx']}, RX={pins['rx']}")

    cocotb.log.info(f"[TEST] Start multi_uart_test")
    
    # Test UART0 - receive "Hello\n"
    msg0 = await uarts[0].get_line()
    cocotb.log.info(f"[TEST] UART0 received: {msg0}")
    # assert msg0 == "Hello\n", f"UART0 expected 'Hello\\n', got '{msg0}'"
    
    # Test UART1 - receive "World"
    msg1 = (await uarts[1].get_line()).rstrip("\r\n")
    cocotb.log.info(f"[TEST] UART1 received: {msg1}")
    # assert msg1 == "World", f"UART1 expected 'World', got '{msg1}'"
    
    # Test UART2 - receive "Test2"
    msg2 = (await uarts[2].get_line()).rstrip("\r\n")
    cocotb.log.info(f"[TEST] UART2 received: {msg2}")
    # assert msg2 == "Test2", f"UART2 expected 'Test2', got '{msg2}'"
    
    # Test UART3 - receive "Test3"
    msg3 = (await uarts[3].get_line()).rstrip("\r\n")
    cocotb.log.info(f"[TEST] UART3 received: {msg3}")
    # assert msg3 == "Test3", f"UART3 expected 'Test3', got '{msg3}'"
    
    # Test UART4 - receive "Test4"
    msg4 = (await uarts[4].get_line()).rstrip("\r\n")
    cocotb.log.info(f"[TEST] UART4 received: {msg4}")
    # assert msg4 == "Test4", f"UART4 expected 'Test4', got '{msg4}'"
    
    # Test UART5 - receive "Test5"
    msg5 = (await uarts[5].get_line()).rstrip("\r\n")
    cocotb.log.info(f"[TEST] UART5 received: {msg5}")
    # assert msg5 == "Test5", f"UART5 expected 'Test5', got '{msg5}'"
    
    # Test UART6 - receive "Test6"
    msg6 = (await uarts[6].get_line()).rstrip("\r\n")
    cocotb.log.info(f"[TEST] UART6 received: {msg6}")
    # assert msg6 == "Test6", f"UART6 expected 'Test6', got '{msg6}'"
    
    # Test UART7 - receive "Test7"
    msg7 = (await uarts[7].get_line()).rstrip("\r\n")
    cocotb.log.info(f"[TEST] UART7 received: {msg7}")
    # assert msg7 == "Test7", f"UART7 expected 'Test7', got '{msg7}'"
    
    cocotb.log.info(f"[TEST] All 8 UARTs tested successfully!")
    cocotb.log.info(f"[TEST] End multi_uart_test") 